mitodl/open-discussions
# Changed HTTP response error handling to behave like form validation

#### Pre-Flight checklist


#### What are the relevant tickets?
Fixes #1129 
Fixes #1130 

#### What's this PR do?
- Changes our form component (HOC) to accept some optional functions that detect HTTP response errors and 'coerce' them to form validation errors
  - Example case: a user enters an unrecognized email in a form, and the HTTP response indicates that the email doesn't exist. You can now define a function that inspects that response and returns an object mapping that error message to the 'email' field of the form
- Updates several components to define these functions
- Removes some behavior that treated HTTP endpoint state as UI state (more on that in the background context section)

#### How should this be manually tested?
Try both of these cases and check to see if the given error shows up like a validation message
- Enter an unrecognized email on the login page (/login), password reset page (/password_reset) and the add moderator/contributor pages (/manage/c/edit/<channel_name>/members/moderators/)
- Enter a valid but incorrect password for a user with email auth

Also try that second case (incorrect password), then click back then forward in your browser. You should no longer see the incorrect password message. You should also be able to enter an email on the first login page with no issues.

#### Where should the reviewer start?
`static/js/hoc/withForm.js` to understand the function props that were changed/added

#### Any background context you want to provide?
The core of the problem being addressed in this PR is that we have been treating HTTP responses as UI state (#1129 has a short thread about this). That presented a few problems for us. If the last request to some endpoint resulted in an error, that really doesn't tell us what we should be showing in the UI. If a form submission results in an error, navigate to a different page, then come back to the same form, the endpoint state will still be the same, but we don't want to continue showing that error any more than we want to show a validation message for a required field that wasn't filled out. Overall, in the context of a form submission, I felt it made the most sense to treat HTTP response errors exactly the same as validation errors.

#### Screenshots (if appropriate)
![ss 2018-09-19 at 12 27 45](https://user-images.githubusercontent.com/14932219/45767166-73e54b80-bc07-11e8-85bc-b1f38da173d1.png)
